### PR TITLE
[util] Add audit feature to compilation DB generator

### PR DESCRIPTION
**Pull request by dmcardle**
**Audit results from commit 0a7a2415abd7193275300474c678ded065529344: https://gist.github.com/dmcardle/b0a11cbce883d846d72e08f8ff9c5d81**

Now, we can audit device CC flags with the following command:

    util/generate_compilation_db.py \
        --audit_device_cc_flags \
        --target //sw/device/... \
        --out cc_flags_audit_compile_commands.json

This new mode restricts the output to source files that are missing required flags. Each object in the emitted JSON file contains a "missing_cc_flags" key that contains exactly which keys are missing.

For example, here's one of the results, with some arguments deleted for clarity:

    {
        "directory": "/home/dan/REPOS/opentitan/bazel-opentitan",
        "arguments": [
            "external/crt/toolchains/lowrisc_rv32imcb/wrappers/clang",
            "--target=riscv32",
            ...
            "sw/device/silicon_creator/rom_ext/keys/fake/sigverify_rsa_keys_fake.c",
            "-o",
            "bazel-out/k8-fastbuild/bin/sw/device/silicon_creator/rom_ext/keys/fake/_objs/fake/sigverify_rsa_keys_fake.o"
        ],
        "file": "sw/device/silicon_creator/rom_ext/keys/fake/sigverify_rsa_keys_fake.c",
        "missing_cc_flags": [
            "-Winvalid-pch",
            "-Wstrict-prototypes",
            "-Wimplicit-fallthrough",
            "-Wextra",
            "-Wgnu",
            "-Wall",
            "-Wno-gnu-zero-variadic-macro-arguments",
            "-Wswitch-default",
            "-Wno-covered-switch-default",
            "-Wno-missing-field-initializers",
            "-Wno-sign-compare",
            "-Wno-error=unused-function",
            "-Wno-unused-parameter"
        ]
    },

Issue #12553

### CODE REVIEW

File: util/generate_compilation_db.py [hunk 1]
```diff
@@ -194,12 +194,60 @@ def build_compile_commands(
...
+        "-Winvalid-pch",
+        "-Wno-covered-switch-default",
+        "-Wno-error=unused-function",
+        "-Wno-gnu-zero-variadic-macro-arguments",
+        "-Wno-missing-field-initializers",
+        "-Wno-sign-compare",
+        "-Wno-unused-parameter",
+        "-Wstrict-prototypes",
+        "-Wswitch-default",
+        "-Wconversion",
```
Comment (alphan) [inline util/generate_compilation_db.py:213]: thanks for including `-Wconversion` too!

File: util/generate_compilation_db.py [hunk 2]
```diff
@@ -194,12 +194,60 @@ def build_compile_commands(
...
+            source_file = command["file"]
+            assert "-march=rv32imc" in arguments
+
+            missing_flags = REQUIRED_DEVICE_CC_FLAGS - set(arguments)
+            if not missing_flags:
+                continue
+
+            troubled_sources.append(source_file)
+
+            command["missing_cc_flags"] = sorted(list(missing_flags))
```
Comment (alphan) [inline util/generate_compilation_db.py:245]: thanks for `sorted(list(...))`

**Review by alphan — APPROVED** [review]
Thanks @dmcardle !

File: util/generate_compilation_db.py (hunk 2 above)
Comment (dmcardle) [reply util/generate_compilation_db.py:245]: Now that you mentioned it, I realize I should update the gist linked at the top of the PR description.

Comment (dmcardle) [reply util/generate_compilation_db.py:245]: Done.

**Pull request merged**